### ISSUE TITLE
Switch to gradle

### ISSUE THREAD

**Issue by kyleboon** [CodeNarc/CodeNarc#39]
I'm wondering if there is interest in switching from Maven to Gradle to build CodeNarc? Gradle has a lot of advantages over Maven plus it integrates more tightly with the groovy ecosystem. If there's interest I'd be willing to help convert the build.

**Comment by topr** [comment]
Guys, I've made an observation. There is a goal - switch to Gradle and then discussion goes totally side track about: site plugin is needed to be written first. My question: is it for sure? My out-of-scope-detector lights orange :bowtie:

I must admit I haven't gone through the build configuration yet to see how such idea would've worked but how about not blocking ourselves with the site part? Just start switching everything apart the site to Gradle while keep shrinking Maven build so finally it would only be responsible only for (legacy at this point) site publication?

I guess this way any contributor may take advantage of Gradle while site is being published only by one person - @chrismair, am I right? Not big overhead then with multi-tooling. We still may expect from contributor to have only one build tool only (or not at if repo-contained Gradle wrapper used). 
This possibly could be reflected within repo structure as well (if it's not yet) where maven site part would've been kept in subfolder and thus slightly hidden 'under the hood'.
This way work with Gradle may be carried on in non-blocking manner while site migration may or may not happen in parallel/later.

BTW. I see sourceforge has empty areas like mailing lists and bug tracker. As communication goes on here, why not to move site into GitHub pages as well and keep everything in one place?
I'm guessing that's quite chunky bite of work but wouldn't it be a good goal to consider on the roadmap?

**Comment by chrismair** [comment]
Thanks for the feedback. Yes, I agree that approach is worth considering (leave site generation in Maven). I'll try to play around with that. Of course, that means I have to get knee-deep in Maven again (ugh :stuck_out_tongue_closed_eyes:). And yes, I see the *irony* of delaying the migration to Gradle because I don't like working with Maven.

**Comment by chrismair** [comment]
Well, that was easier than I feared. I have created a site-only Maven POM file, and that seems to work pretty well.

**Comment by topr** [comment]
Good to hear that. I've done basic gradle setup. Build works well. Tests are green.
I'm going to push it today. I hope you can tell what are the missing bits :)

**Comment by chrismair** [comment]
Thanks very much. I have created a **gradle-migration** branch.

Other than the items you listed, we will also need to add support to create the .zip and tar.gz distribution files.

**Comment by topr** [comment]
The same PR against the new branch: https://github.com/CodeNarc/CodeNarc/pull/96
You're welcome to add comments in case of any questions or points.

What the distro packages should contain? Are they deployed to maven repo as well?

**Comment by chrismair** [comment]
Thanks Tomasz.

The distribution files are not deployed to Maven, just hosted on Sourceforge. They contain the CodeNarc jar, site files, dependent jars (Log4J) and source. See:
https://sourceforge.net/projects/codenarc/files/codenarc/CodeNarc%200.23/

That being said, I am open to considering a different approach. What really is necessary/useful to be downloaded?

**Comment by topr** [comment]
I'm going to raise PR with maven basic maven deployment soon.
BTW. What's the maven repo URL, Shoudl I just use maven central? Is there need of providing credentials for deployment?

Not sure what's the exact need for having distribution packages as I rather use tools like CodeNarc (and any libs to be honest as well) via dependency management tool (Maven, Gradle).

I don't mind preparing relevant Gradle configuration for creating distros though. We may postpone that as less important or skip at all if you think it may be not useful enough.
However maybe there are still some folks who prefer the ol' good way of downloading and unzipping by hand so we may keep that for the sake of people habits 'backward compatibility'.

By the way, I raised new issue https://github.com/CodeNarc/CodeNarc/issues/100
and a PR https://github.com/CodeNarc/CodeNarc/pull/99 to sync up **gradle-migration** branch with master. Please give your thoughts about the 1st one and the 2nd as well (or merge straight away if you think that's all right).

Thanks for merging the previous one.

**Comment by chrismair** [comment]
Re: Maven repo:
I currently use the Maven Release Plugin to publish to:
https://oss.sonatype.org/index.html#welcome

Steps are something like:
1.	mvn deploy
2.	mvn release:clean
3.	mvn release:prepare
4.	mvn release:perform 
5.	Go to http://oss.sonatype.org/index.html#welcome and log in.
	- Select Staging Repositories
	- Select org.codenarc and click [Close]. 
	- Download and verify artifacts (jar) from the staged repository (Under the Content tab).
	- If problems are discovered, you can right-click on the staged repository and select [Drop].
	- If they are ready to be synced to Central, right click on the staged org.codenarc repository and select [Release].

This may take some wrangling.

**Comment by chrismair** [comment]
I have checked in changes to **build.gradle** (on the gradle-migration branch), publish the artifact jars to Sonatype (OSSRH) Maven Repository. This is based on the instructions (well) documented at:
http://central.sonatype.org/pages/gradle.html and also http://yennicktrevels.com/blog/2013/10/11/automated-gradle-project-deployment-to-sonatype-oss-repository/. 

I checked that the snapshot artifacts were written to  https://oss.sonatype.org/content/repositories/snapshots/.

That was much less trouble than I expected.

**Comment by topr** [comment]
That's a great news :+1:

Some time ago it was more hassle and buggy as well. They've got serious about development over there at Gradle's and it's pretty nice now.

I was keen to try the new _maven-publish_ plugin. I guess, publication with 'maven' plugin might get decommissioned in some time. It will take ages till then though, so lets move along. I got distracted by issues anyway:
1. Javadoc generation gives hard time with java8. I didn't want to introduce error for deployment done on java8. You've got it there so I assume you have older jdk in use. I've got solution for this one already (not rebased with your change though). There is a PR to overcome the problem https://github.com/CodeNarc/CodeNarc/pull/103
2. I've noticed that only for java sources docs are being generated. I'm afraid this might be the case for the config you added as well. Or is it perhaps desired state? If no, what would be the layout of docs: two separate jars for groovydocs and javadocs or one somehow combined?
3. New one: gradle build fails now with any task execution as `ossrhUsername` and `ossrhPassword` are undefined. Do you have them set anyhow? Here's pull request, perhaps you might like to pass them as command line arguments or set in local ~/.gradle/gradle.properties - both supported: https://github.com/CodeNarc/CodeNarc/pull/104

**Comment by chrismair** [comment]
Thanks for the enhancements.

re: (2) Yeah, on the Maven build, it was only generating Javadocs for the Java files. The Javadoc/Groovydoc was always problematic. It would be nice to solve that at some point, but I am not going to let that be a showstopper.

**Comment by topr** [comment]
There is task for generating groovy docs already added by groovy plugin. Adding another artifact with groovydoc is simple so we may do it straight away. Docs won't be combined though. Still better than not publishing them at all, I guess.

By the way, is there any specific reason that some of the sources are written in java and some are in groovy?

Just not to miss this one, I gave some more explanation under the closed PR https://github.com/CodeNarc/CodeNarc/pull/104#issuecomment-108275942

**Comment by chrismair** [comment]
Some of the source files were converted to Java for performance -- it made a significant improvement in run times.

The Maven POM for the site is in the `site-pom.xml` file. My intent was just to delete pom.xml when we made the switch. I think I prefer have a special-named "site-pom.xml" to make it clear that it is only for the site -- I want to avoid someone checking out the project and seeing a pom.xml and assuming it is a *regular* Maven project.

For now I think I'll just leave the two site-specific files in the root folder.

**Comment by topr** [comment]
>I want to avoid someone checking out the project and seeing a pom.xml and assuming it is a regular Maven project.
That was my intention as well when I proposed moving it to sub-folder. Having `site-pom.xml` is kind of the same :+1:

I wonder whether Groovy static compilation would've given the same performance improvement as conversion to Java. Just a thought.

Anyway, anything missing before migration branch is going to be merged into master?

**Comment by chrismair** [comment]
> I wonder whether Groovy static compilation would've given the same performance 
> improvement as conversion to Java.
We (actually Hamlet) made those optimizations while still on Groovy 1.7.

**Comment by chrismair** [comment]
I have removed the pom.xml file from the gradle-migration branch.

I think I am going to take the lazy option and not worry about building the zip/tar.gz distribution files. I'll just publish the jar files and see if anyone complains.

So, unless there are objections, I think this Gradle migration is good enough, and ready to be merged.

**Comment by chrismair** [comment]
I have merged in the **gradle-migration** branch. Thanks very much Tomasz, Marcin and Kyle for your contributions, help and consultation.

**Comment by topr** [comment]
Always glad to contribute.
I'm curious, have you released with Gradle yet? Does all work as it should?

BTW. adding support for traits may a good reason to release and publish new version :)
